File: Leetcode/Easy/20. Valid Parentheses/solution.php
class Solution {

    /**
     * @param String $s
     * @return Boolean
     */
    function isValid($s) {
        $stack = [];
        $map = [
            ")" => "(",
            "]" => "[",
            "}" => "{",
        ];
        for($i = 0; $i < strlen($s); $i++){
            $char = $s[$i];
            if(array_key_exists($char, $map)){
                if(empty($stack) || $stack[count($stack) - 1] != $map[$char]){
                    return false;
                }
                array_pop($stack);
            } else {
                array_push($stack, $char);
            }
        }
        return empty($stack);
    }
}

File: Leetcode/Easy/20. Valid Parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashMap = {
            "}": "{",
            "]": "[",
            ")": "(",
        }
        stack = []
        for c in s:
            if c not in hashMap:
                stack.append(c)
                continue
            if not stack or stack[-1] != hashMap[c]:
                return False
            stack.pop()
        return not stack
